Core/src/org/sleuthkit/autopsy/healthmonitor/ServicesHealthMonitor.java

```diff
@@ -0,0 +1,767 @@
+        if(isEnabled.get() && (metric != null)) {
+            metric.stopTiming();
+            try {
+                getInstance().addTimingMetric(metric);
+            } catch (HealthMonitorException ex) {
+                // We don't want calling methods to have to check for exceptions, so just log it
+                logger.log(Level.SEVERE, "Error adding timing metric", ex);
+            }
+        }
+    }
+    
+    /**
+     * Add the timing metric data to the map.
+     * @param metric The metric to add. stopTiming() should already have been called.
+     */
+    private void addTimingMetric(TimingMetric metric) throws HealthMonitorException {
+
+        // Do as little as possible within the synchronized block to minimize
+        // blocking with multiple threads.
+        synchronized(this) {
+            // There's a small check-then-act situation here where isEnabled
+            // may have changed before reaching this code. This is fine - 
+            // the map still exists and any extra data added after the monitor
+            // is disabled will be deleted if the monitor is re-enabled. This
+            // seems preferable to doing another check on isEnabled within
+            // the synchronized block.
+            if(timingInfoMap.containsKey(metric.getName())) {
+                timingInfoMap.get(metric.getName()).addMetric(metric);
+            } else {
+                timingInfoMap.put(metric.getName(), new TimingInfo(metric));
+            }
+        }
+    }
+    
+    /**
+     * Write the collected metrics to the database.
+     * @throws HealthMonitorException 
+     */
+    private void writeCurrentStateToDatabase() throws HealthMonitorException {
+                
+        Map<String, TimingInfo> timingMapCopy;
+        
+        // Do as little as possible within the synchronized block since it will
+        // block threads attempting to record metrics.
+        synchronized(this) {
+            if(! isEnabled.get()) {
+                return;
+            }
+            
+            // Make a shallow copy of the timing map. The map should be small - one entry
+            // per metric name.
+            timingMapCopy = new HashMap<>(timingInfoMap);
+            timingInfoMap.clear();
+        }
+        logger.log(Level.INFO, "Writing health monitor metrics to database");
+
+        String hostName;
+        try {
+            hostName = java.net.InetAddress.getLocalHost().getHostName();
+        } catch (java.net.UnknownHostException ex) {
+            // Write it to the database but log a warning
+            logger.log(Level.WARNING, "Unable to look up host name");
+            hostName = "unknown";
+        }
+        
+        // Check if there's anything to report (right now we only have the timing map)
+        if(timingMapCopy.keySet().isEmpty()) {
+            return;
+        }
+        
+        // Write to the database
+        CoordinationService.Lock lock = getSharedDbLock();
```

Comment:
I don't quite get the need for database locking here. Given that this is a multi-user/PostgreSQL only thing is database locking not sufficient? If 2 different processes attempt to insert data with the same name and host I'm guessing they will both succeed (since neither of those fields are key fields in the schema below).

Reply:
Yes inserting data normally by multiple nodes should be fine - that's why it's a shared lock. But eventually I'm sure we'll have to update the database schema during which time we'll need to stop the inserts by getting the exclusive lock.